com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/context.py

```diff
@@ -158,82 +57,199 @@ def _print_process_stderr(process, fn):
-        stderr_thread.daemon = True            
-        stderr_thread.start()
-        if not ctxtype_was == "JUPYTER":
-            stdout_thread = threading.Thread(target=_print_process_stdout, args=([process]))
-            stdout_thread.daemon = True
-            stdout_thread.start()                
-            process.wait()
-            return None
-        else:            
-            return process.stdout
-    except:
-        _print_exception("Error starting java subprocess for submission")
-        raise
+def submit(ctxtype, app_topology, config = None, username = None, password = None, rest_api_url = None, log_level=logging.INFO):
+    """
+    Submits a topology with the specified context type.
+    
+    Args:
+        ctxtype (string): context type.  Values include:
+        * DISTRIBUTED - the topology is submitted to a Streams instance.
+          The bundle is submitted using `streamtool` which must be setup to submit without requiring authentication input.
+        * STANDALONE - the topology is executed directly as an Streams standalone application.
+          The standalone execution is spawned as a separate process
+        * BUNDLE - execution of the topology produces an SPL application bundle
+          (.sab file) that can be submitted to an IBM Streams instance as a distributed application.
+        * JUPYTER - the topology is run in standalone mode, and context.submit returns a stdout streams of bytes which 
+          can be read from to visualize the output of the application.
+        app_topology: a Topology object or Topology.graph object
+        
+    Returns:
+        An output stream of bytes if submitting with JUPYTER, otherwise returns None.
+    """    
+    logger.setLevel(log_level)
+    context_submitter = ContextSubmitter(app_topology, config, username, password, rest_api_url)
+    context_submitter.submit(ctxtype)
+
+class ContextSubmitter:
+    """
+    ContextSubmitter:
+        Responsible for performing the correct submission depending on a number of factors, including: the
+        presence/absence of a streams install, the type of context, and whether the user seeks to retrieve data via rest.
+    """
+
+    def __init__(self, app_topology, config = None, username = None, password = None, rest_api_url = None):
+        self.app_topology = app_topology.graph
+        self.config = config
+        self.username = username
+        self.password = password
+        self.rest_api_url = rest_api_url
+
+        if self.config is None:
+            self.config = {}
+
+        # deserialize vcap config before passing it off to Java
+        if 'topology.service.vcap' in config:
+            config['topology.service.vcap'] = json.loads(config['topology.service.vcap'])
+
+        # encode the relevant python version information into the config
+        self._do_pyversion_initialization()
+
+
+    def _do_pyversion_initialization(self):
```

Comment:
Not sure, exactly. @sdickes added this initialization code as part of the Python 2.7 support.